ramalama/gpu_detector.py

```diff
@@ -147,6 +149,47 @@ def get_macos_gpu(self):
             logging.error(f"Unexpected error while detecting macOS GPU: {e}")
             return [{"GPU": "Unknown", "Error": str(e)}]
 
+    def get_ascend_npu(self):
+        """Detects Ascend NPUs using npu-smi (Linux only)."""
+        if platform.system() != "Linux":
+            return  # Skip on macOS and other platforms
+
+        try:
+            gpus = []
+            # get total npu number
+            total_count_process = subprocess.run(["npu-smi", "info", "-l"], capture_output=True, text=True, check=True)
+            total_count_output = total_count_process.stdout
+            total_count_match = re.search(r"Total Count\s+:\s*(\d+)", total_count_output)
+            if not total_count_match:
+                print("Error: Could not determine total NPU count.")
+                return
+
+            total_count = int(total_count_match.group(1))
+            for npu_id in range(total_count):
+                gpu_info = {"GPU": npu_id, "Vendor":"Ascend", "Env":"CANN_VISIBLE_DEVICES"}
+                # get memory of each card
+                npu_info_process = subprocess.run(["npu-smi", "info", "-t", "memory", "-i", str(npu_id)], capture_output=True, text=True, check=True)
+                npu_info_output = npu_info_process.stdout
+
+                # parse memory
+                hbm_capacity_match = re.search(r"HBM Capacity\(MB\)\s+:\s*(\d+)", npu_info_output)
+                if not hbm_capacity_match:
+                    print(f"Error: Could not find HBM Capacity for NPU {npu_id}.")
+                    continue
+                hbm_capacity = int(hbm_capacity_match.group(1))
+
+                self._update_best_gpu(hbm_capacity, npu_id, "CANN_VISIBLE_DEVICES")
+                gpu_info["VRAM"] = hbm_capacity
+                gpus.append(gpu_info)
+            
+            return gpus
+        except subprocess.CalledProcessError as e:
+            logging.error(f"Failed to detect ascend npu on linux: {e}")
+            raise RuntimeError("`npu-smi` not found. No ASCEND NPU detected or drivers missing.")
```

Comment:
This will cause the error to be reported twice.  The command line tool will display the RuntimeError to the user.